### ISSUE TITLE
Allow operator aliases to be declared for data constructors

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1833]

**Comment by user471** [comment]
Would operator aliases work with pattern matching?

**Comment by cryogenian** [comment]
What about type constructors? e.g. `type a \/ b = Either a b`

**Comment by garyb** [comment]
@user471 maybe, I hadn't considered that. We'd need to do some work on the parser, but it might well be possible.

@cryogenian I'm in favour of that too, but it's a separate issue. I'd like it to work much like it does for value operators, so you can still set associativity and precedence.

This issue is just about tweaking the parser to accept propernames as well as idents.

**Comment by paf31** [comment]
What's the use case here? Shouldn't we be encouraging smart constructors?

**Comment by jdegoes** [comment]
Smart constructors don't provide a solution to pattern matching, e.g.:

```purescript
foo :: forall a b c d. Tuple4 a b c d -> d
foo (a /\ b /\ c /\ d) = d
```

**Comment by paf31** [comment]
I think we should be very careful not to introduce too many opportunities for abusing operators.

**Comment by jdegoes** [comment]
@paf31 I'd argue that a discrepancy between construction and deconstruction leads to code that is inherently difficult to read. Compare:

```purescript
let foo = a /\ b /\ c /\ d
in case foo of
  (Tuple (Tuple (Tuple a b) c) d) -> d
```

with:

```purescript
let foo = a /\ b /\ c /\ d
in case foo of
  a /\ b /\ c /\ d -> d
```

If you allow operators, then operators will be used for constructors. If operators are used for constructors, then it greatly improves readability to allow them for deconstructors.

In addition, I'd suggest that PureScript (the language) should be fairly un-opinionated about whether operators are good or bad, and let that decision live in user-land communities so that every (niche) community can make their own choice about that.

**Comment by paf31** [comment]
> I'd suggest that PureScript (the language) should be fairly un-opinionated about whether operators are good or bad

Sounds reasonable. Let's just be careful to avoid overuse in core libraries then.

**Comment by jdegoes** [comment]
> Sounds reasonable. Let's just be careful to avoid overuse in core libraries then.

:+1:

**Comment by garyb** [comment]
The use case is the `Tuple` operator `(/\)` in `Data.Tuple.Nested`. I thought it was weird I had to declare it via the `tuple2` function rather than being able to just use `Tuple`. But yep, now it's been raised, it seems like a nice way of getting cons patterns back too (albeit with `Nil` instead of `[]`), along with the nested `Tuple` stuff, etc.

**Comment by paf31** [comment]
I'm in favour, but I think we should bump this to 0.9, since there are quite a lot of things which need to change to implement this: parsing, pretty printing, desugaring operators, traversals for new binder types.

**Comment by garyb** [comment]
I was purely going to tweak the parser to make it possible to declare them for the `Tuple`/`tuple2` case I described above for now, but yeah, perhaps it would make sense to do it altogether in 0.9?

**Comment by paf31** [comment]
Well, I suppose let's get the syntax now, and if we can implement the easy case (construction), then that's good. Then we can implement patterns in 0.9.

**Comment by garyb** [comment]
It'll have a look. If it's going to take longer than 15 minutes I'll bump the whole thing to 0.9 :smile:

**Comment by paf31** [comment]
Sounds good, thanks!

**Comment by garyb** [comment]
Bumping to 0.9. It's probably better to do it all at once anyway.

**Comment by garyb** [comment]
Resolved by #1890
